File: girly-app/cycle.go
package main

import (
	"time"
)

const (
	dateLayout = "2006-01-02"
)

func todayStr() string { return time.Now().Format(dateLayout) }

func parseDate(s string) (time.Time, bool) {
	t, err := time.Parse(dateLayout, s)
	return t, err == nil
}

// CycleSummary is the computed prediction model handed to the frontend.
type CycleSummary struct {
	HasData          bool   `json:"has_data"`
	CycleDay         int    `json:"cycle_day"`
	CycleLength      int    `json:"cycle_length"`
	AvgCycleLength   int    `json:"avg_cycle_length"`
	PeriodLength     int    `json:"period_length"`
	CyclesLogged     int    `json:"cycles_logged"`
	Phase            string `json:"phase"`             // menstrual | follicular | ovulatory | luteal
	PhaseLabel       string `json:"phase_label"`       // human label e.g. "Luteal Phase (Free Window)"
	NextPeriodDate   string `json:"next_period_date"`  // YYYY-MM-DD
	DaysUntilPeriod  int    `json:"days_until_period"` // 0 = today
	OvulationDate    string `json:"ovulation_date"`
	FertileStart     string `json:"fertile_start"`
	FertileEnd       string `json:"fertile_end"`
	LastPeriodStart  string `json:"last_period_start"`
	PredictionWindow int    `json:"prediction_window"` // ± days
	Tip              string `json:"tip"`               // gentle wellness tip for the current phase
}

// ComputeCycle derives predictions from a user's period history.
func ComputeCycle(u User) CycleSummary {
	sum := CycleSummary{
		CycleLength:      28,
		AvgCycleLength:   28,
		PeriodLength:     u.PeriodLength,
		CyclesLogged:     len(u.PeriodStarts),
		PredictionWindow: 2,
	}
	if sum.PeriodLength <= 0 {
		sum.PeriodLength = 5
	}

	// Average cycle length from completed cycles (gaps between starts),
	// clamped to a medically plausible 21–45 day range.
	if len(u.PeriodStarts) >= 2 {
		total := 0
		n := 0
		for i := 1; i < len(u.PeriodStarts); i++ {
			a, ok1 := parseDate(u.PeriodStarts[i-1])
			b, ok2 := parseDate(u.PeriodStarts[i])
			if ok1 && ok2 {
				gap := int(b.Sub(a).Hours() / 24)
				if gap >= 15 && gap <= 60 {
					total += gap
					n++
				}
			}
		}
		if n > 0 {
			avg := total / n
			if avg < 21 {
				avg = 21
			}
			if avg > 45 {
				avg = 45
			}
			sum.AvgCycleLength = avg
			sum.CycleLength = avg
		}
	}

	if len(u.PeriodStarts) == 0 {
		sum.Phase, sum.PhaseLabel = "learn", "Learn Mode"
		sum.Tip = "Your journey is just beginning — explore the Learn tab for friendly guides."
		return sum
	}

	last, ok := parseDate(u.PeriodStarts[len(u.PeriodStarts)-1])
	if !ok {
		return sum
	}
	sum.HasData = true
	sum.LastPeriodStart = last.Format(dateLayout)

	// "Today" as a clean UTC-midnight instant matching the local calendar
	// date, so every comparison below stays day-level and timezone-safe.
	now, _ := parseDate(todayStr())
	// Roll the reference start forward so today always falls inside a cycle.
	start := last
	for start.AddDate(0, 0, sum.AvgCycleLength).Before(now) || start.AddDate(0, 0, sum.AvgCycleLength).Equal(now) {
		start = start.AddDate(0, 0, sum.AvgCycleLength)
	}
	cycleDay := int(now.Sub(start).Hours()/24) + 1
	if cycleDay < 1 {
		cycleDay = 1
	}
	sum.CycleDay = cycleDay

	next := start.AddDate(0, 0, sum.AvgCycleLength)
	sum.NextPeriodDate = next.Format(dateLayout)
	sum.DaysUntilPeriod = int(next.Sub(now).Hours() / 24)

	// Ovulation ~14 days before the next period; fertile window spans
	// ovulation −5 … ovulation +1.
	ovu := next.AddDate(0, 0, -14)
	sum.OvulationDate = ovu.Format(dateLayout)
	sum.FertileStart = ovu.AddDate(0, 0, -5).Format(dateLayout)
	sum.FertileEnd = ovu.AddDate(0, 0, 1).Format(dateLayout)

	sum.Phase, sum.PhaseLabel = phaseFor(now, cycleDay, sum.PeriodLength, ovu)
	sum.Tip = tipFor(sum.Phase, sum.DaysUntilPeriod)
	return sum
}

func phaseFor(now time.Time, cycleDay, periodLen int, ovu time.Time) (string, string) {
	fertileStart := ovu.AddDate(0, 0, -5)
	fertileEnd := ovu.AddDate(0, 0, 1)

	switch {
	case cycleDay <= periodLen:
		return "menstrual", "Menstrual Phase"
	case !now.Before(fertileStart) && !now.After(fertileEnd):
		if now.Equal(ovu) {
			return "ovulatory", "Ovulation Peak"
		}
		return "ovulatory", "Ovulatory Phase (Fertile Window)"
	case now.Before(fertileStart):
		return "follicular", "Follicular Phase (Rising Energy)"
	default:
		return "luteal", "Luteal Phase (Free Window)"
	}
}

func tipFor(phase string, daysUntil int) string {
	switch phase {
	case "menstrual":
		return "Rest is productive today. A warm compress, iron-rich foods, and extra hydration help your body replenish."
	case "follicular":
		return "Energy is on the rise — a lovely window for movement, projects, and trying something new."
	case "ovulatory":
		return "You may feel outgoing and confident. Great days for social plans and energizing workouts."
	default:
		if daysUntil <= 4 {
			return "Your body is naturally winding down. Drink warm chamomile tea, hydrate, and prioritize rest today."
		}
		return "Steady self-care now softens the week ahead: magnesium-rich snacks and gentle stretching go a long way."
	}
}

// DayStatus describes one calendar day for the tracker calendar UI.
type DayStatus struct {
	Date        string `json:"date"`
	Day         int    `json:"day"`
	Phase       string `json:"phase,omitempty"`       // menstrual|follicular|ovulatory|luteal (inferred/predicted)
	Kind        string `json:"kind"`                  // logged_period|predicted|fertile|ovulation|plain
	IsToday     bool   `json:"is_today"`
	HasLog      bool   `json:"has_log"`
	IsWindow    bool   `json:"is_window"` // inside ±2d prediction window
}

// BuildCalendar produces the day grid (with leading blanks as empty Date strings)
// for a given month, colored by logs and predictions.
func BuildCalendar(u User, sum CycleSummary, month time.Time) []DayStatus {
	first := time.Date(month.Year(), month.Month(), 1, 0, 0, 0, 0, time.UTC)
	daysInMonth := time.Date(month.Year(), month.Month()+1, 0, 0, 0, 0, 0, time.UTC).Day()

	logsByDate := map[string]DayLog{}
	for _, l := range u.Logs {
		logsByDate[l.Date] = l
	}
	periodDays := map[string]bool{}
	for _, start := range u.PeriodStarts {
		if t, ok := parseDate(start); ok {
			pl := u.PeriodLength
			if pl <= 0 {
				pl = 5
			}
			for i := 0; i < pl; i++ {
				periodDays[t.AddDate(0, 0, i).Format(dateLayout)] = true
			}
		}
	}

	// Projected period starts: project the average cycle from the last known
	// start, forward and backward, covering the requested month. The bound
	// extends past month end by the prediction window so a start early next
	// month still colors its ±2d window days that spill into this month.
	monthEnd := first.AddDate(0, 1, 0).AddDate(0, 0, sum.PredictionWindow+7)

	predicted := map[string]bool{}
	window := map[string]bool{}
	if sum.HasData {
		if last, ok := parseDate(u.PeriodStarts[len(u.PeriodStarts)-1]); ok {
			for base := last.AddDate(0, 0, -sum.AvgCycleLength*4); !base.After(monthEnd); base = base.AddDate(0, 0, sum.AvgCycleLength) {
				for i := 0; i < sum.PeriodLength; i++ {
					d := base.AddDate(0, 0, i)
					if !periodDays[d.Format(dateLayout)] {
						predicted[d.Format(dateLayout)] = true
					}
				}
				for w := -sum.PredictionWindow; w <= sum.PredictionWindow; w++ {
					wd := base.AddDate(0, 0, w)
					if !periodDays[wd.Format(dateLayout)] {
						window[wd.Format(dateLayout)] = true
					}
				}
			}
		}
	}

	// Fertile / ovulation days across projected cycles.
	fertile := map[string]bool{}
	ovuDays := map[string]bool{}
	if sum.HasData {
		if last, ok := parseDate(u.PeriodStarts[len(u.PeriodStarts)-1]); ok {
			for base := last.AddDate(0, 0, -sum.AvgCycleLength*4); !base.After(monthEnd); base = base.AddDate(0, 0, sum.AvgCycleLength) {
				ovu := base.AddDate(0, 0, sum.AvgCycleLength-14)
				for i := -5; i <= 1; i++ {
					fertile[ovu.AddDate(0, 0, i).Format(dateLayout)] = true
				}
				ovuDays[ovu.Format(dateLayout)] = true
			}
		}
	}

	today := todayStr()
	var out []DayStatus
	// weekday offset — calendar starts on Monday
	offset := (int(first.Weekday()) + 6) % 7
	for i := 0; i < offset; i++ {
		out = append(out, DayStatus{})
	}
	for d := 1; d <= daysInMonth; d++ {
		date := time.Date(month.Year(), month.Month(), d, 0, 0, 0, 0, time.UTC)
		ds := date.Format(dateLayout)
		status := DayStatus{Date: ds, Day: d, IsToday: ds == today, HasLog: logsByDate[ds].Date != ""}

		switch {
		case periodDays[ds]:
			status.Kind = "logged_period"
			status.Phase = "menstrual"
		case ovuDays[ds]:
			status.Kind = "ovulation"
			status.Phase = "ovulatory"
		case fertile[ds]:
			status.Kind = "fertile"
			status.Phase = "ovulatory"
		case predicted[ds] || window[ds]:
			status.Kind = "predicted"
			status.Phase = "menstrual"
			status.IsWindow = window[ds]
		case sum.HasData:
			status.Kind = "plain"
		}
		out = append(out, status)
	}
	return out
}

File: girly-app/cycle.py
"""Girly 🌸 — cycle-day / phase / fertile-window prediction math.

Python port of cycle.go. Users are plain dicts (as loaded from the JSON
store); all prediction logic is day-level and timezone-safe.
"""

import calendar
from datetime import date, datetime, timedelta

DATE_FMT = "%Y-%m-%d"


def today_str() -> str:
    return date.today().strftime(DATE_FMT)


def parse_date(s):
    """Parse a YYYY-MM-DD string; returns a date or None."""
    try:
        return date.fromisoformat(s)
    except (TypeError, ValueError):
        return None


def parse_month(s):
    """Parse a YYYY-MM string; returns a date (first of month) or None."""
    try:
        return datetime.strptime(s, "%Y-%m").date()
    except (TypeError, ValueError):
        return None


def compute_cycle(user):
    """Derive the prediction summary handed to the frontend."""
    starts = user.get("period_starts") or []
    period_length = user.get("period_length") or 0
    summary = {
        "has_data": False,
        "cycle_day": 0,
        "cycle_length": 28,
        "avg_cycle_length": 28,
        "period_length": period_length if period_length > 0 else 5,
        "cycles_logged": len(starts),
        "phase": "",
        "phase_label": "",
        "next_period_date": "",
        "days_until_period": 0,
        "ovulation_date": "",
        "fertile_start": "",
        "fertile_end": "",
        "last_period_start": "",
        "prediction_window": 2,
        "tip": "",
    }

    # Average cycle length from completed cycles (gaps between starts),
    # clamped to a medically plausible 21–45 day range.
    if len(starts) >= 2:
        total = n = 0
        for prev, cur in zip(starts, starts[1:]):
            a, b = parse_date(prev), parse_date(cur)
            if a and b:
                gap = (b - a).days
                if 15 <= gap <= 60:
                    total += gap
                    n += 1
        if n > 0:
            avg = total // n
            avg = max(21, min(45, avg))
            summary["avg_cycle_length"] = avg
            summary["cycle_length"] = avg

    if not starts:
        summary["phase"] = "learn"
        summary["phase_label"] = "Learn Mode"
        summary["tip"] = (
            "Your journey is just beginning — explore the Learn tab for friendly guides."
        )
        return summary

    last = parse_date(starts[-1])
    if last is None:
        return summary
    summary["has_data"] = True
    summary["last_period_start"] = last.strftime(DATE_FMT)

    # "Today" as a clean calendar date, so every comparison below stays
    # day-level and timezone-safe.
    now = parse_date(today_str())
    avg = summary["avg_cycle_length"]
    # Roll the reference start forward so today always falls inside a cycle.
    start = last
    while start + timedelta(days=avg) <= now:
        start = start + timedelta(days=avg)
    cycle_day = (now - start).days + 1
    summary["cycle_day"] = max(1, cycle_day)

    next_start = start + timedelta(days=avg)
    summary["next_period_date"] = next_start.strftime(DATE_FMT)
    summary["days_until_period"] = (next_start - now).days

    # Ovulation ~14 days before the next period; fertile window spans
    # ovulation −5 … ovulation +1.
    ovu = next_start - timedelta(days=14)
    summary["ovulation_date"] = ovu.strftime(DATE_FMT)
    summary["fertile_start"] = (ovu - timedelta(days=5)).strftime(DATE_FMT)
    summary["fertile_end"] = (ovu + timedelta(days=1)).strftime(DATE_FMT)

    summary["phase"], summary["phase_label"] = _phase_for(
        now, cycle_day, summary["period_length"], ovu
    )
    summary["tip"] = _tip_for(summary["phase"], summary["days_until_period"])
    return summary


def _phase_for(now, cycle_day, period_len, ovu):
    fertile_start = ovu - timedelta(days=5)
    fertile_end = ovu + timedelta(days=1)

    if cycle_day <= period_len:
        return "menstrual", "Menstrual Phase"
    if fertile_start <= now <= fertile_end:
        if now == ovu:
            return "ovulatory", "Ovulation Peak"
        return "ovulatory", "Ovulatory Phase (Fertile Window)"
    if now < fertile_start:
        return "follicular", "Follicular Phase (Rising Energy)"
    return "luteal", "Luteal Phase (Free Window)"


def _tip_for(phase, days_until):
    if phase == "menstrual":
        return (
            "Rest is productive today. A warm compress, iron-rich foods, and extra "
            "hydration help your body replenish."
        )
    if phase == "follicular":
        return (
            "Energy is on the rise — a lovely window for movement, projects, and "
            "trying something new."
        )
    if phase == "ovulatory":
        return "You may feel outgoing and confident. Great days for social plans and energizing workouts."
    if days_until <= 4:
        return (
            "Your body is naturally winding down. Drink warm chamomile tea, hydrate, "
            "and prioritize rest today."
        )
    return (
        "Steady self-care now softens the week ahead: magnesium-rich snacks and "
        "gentle stretching go a long way."
    )


def _blank_day():
    return {
        "date": "",
        "day": 0,
        "kind": "",
        "is_today": False,
        "has_log": False,
        "is_window": False,
    }


def build_calendar(user, summary, month):
    """Produce the day grid (with leading blanks as empty-date entries) for a
    given month, colored by logs and predictions."""
    first = month.replace(day=1)
    days_in_month = calendar.monthrange(month.year, month.month)[1]

    logs_by_date = {log["date"]: log for log in (user.get("logs") or [])}
    period_days = set()
    period_len = user.get("period_length") or 0
    if period_len <= 0:
        period_len = 5
    for start in user.get("period_starts") or []:
        t = parse_date(start)
        if t:
            for i in range(period_len):
                period_days.add((t + timedelta(days=i)).strftime(DATE_FMT))

    # Projected period starts: project the average cycle from the last known
    # start, forward and backward, covering the requested month. The bound
    # extends past month end by the prediction window so a start early next
    # month still colors its ±2d window days that spill into this month.
    month_end = (first + timedelta(days=32)).replace(day=1) + timedelta(
        days=summary["prediction_window"] + 7
    )

    avg = summary["avg_cycle_length"]
    starts = user.get("period_starts") or []
    predicted, window = set(), set()
    fertile, ovu_days = set(), set()
    if summary["has_data"] and starts:
        last = parse_date(starts[-1])
        if last:
            base = last - timedelta(days=avg * 4)
            while base <= month_end:
                for i in range(summary["period_length"]):
                    d = (base + timedelta(days=i)).strftime(DATE_FMT)
                    if d not in period_days:
                        predicted.add(d)
                for w in range(-summary["prediction_window"], summary["prediction_window"] + 1):
                    wd = (base + timedelta(days=w)).strftime(DATE_FMT)
                    if wd not in period_days:
                        window.add(wd)
                # Fertile / ovulation days across projected cycles.
                ovu = base + timedelta(days=avg - 14)
                for i in range(-5, 2):
                    fertile.add((ovu + timedelta(days=i)).strftime(DATE_FMT))
                ovu_days.add(ovu.strftime(DATE_FMT))
                base = base + timedelta(days=avg)

    today = today_str()
    out = []
    # weekday offset — calendar starts on Monday (Python: Monday == 0)
    for _ in range(first.weekday()):
        out.append(_blank_day())
    for d in range(1, days_in_month + 1):
        ds = first.replace(day=d).strftime(DATE_FMT)
        status = _blank_day()
        status.update(
            {"date": ds, "day": d, "is_today": ds == today, "has_log": ds in logs_by_date}
        )

        if ds in period_days:
            status["kind"] = "logged_period"
            status["phase"] = "menstrual"
        elif ds in ovu_days:
            status["kind"] = "ovulation"
            status["phase"] = "ovulatory"
        elif ds in fertile:
            status["kind"] = "fertile"
            status["phase"] = "ovulatory"
        elif ds in predicted or ds in window:
            status["kind"] = "predicted"
            status["phase"] = "menstrual"
            status["is_window"] = ds in window
        elif summary["has_data"]:
            status["kind"] = "plain"
        out.append(status)
    return out
